File: app/src/main/java/com/example/ruslanyussupov/popularmovies/detail/DetailViewModel.java
package com.example.ruslanyussupov.popularmovies.detail;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.example.ruslanyussupov.popularmovies.App;
import com.example.ruslanyussupov.popularmovies.IoScheduler;
import com.example.ruslanyussupov.popularmovies.Result;
import com.example.ruslanyussupov.popularmovies.Utils;
import com.example.ruslanyussupov.popularmovies.data.DataSource;
import com.example.ruslanyussupov.popularmovies.data.model.Movie;
import com.example.ruslanyussupov.popularmovies.data.model.Review;
import com.example.ruslanyussupov.popularmovies.data.model.Video;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;

public class DetailViewModel extends ViewModel {

    @Inject
    DataSource dataSource;

    @Inject
    IoScheduler ioScheduler;

    @Inject
    Utils utils;

    private final Movie movie;
    private final MutableLiveData<Result<List<Video>>> videosResultLiveData;
    private final MutableLiveData<Result<List<Review>>> reviewsResultLiveData;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public DetailViewModel(Movie movie) {
        this.movie = movie;
        App.getComponent().inject(this);
        videosResultLiveData = new MutableLiveData<>();
        reviewsResultLiveData = new MutableLiveData<>();
        subscribeToVideosDataSource();
        subscribeToReviewsDataSource();
    }

    public LiveData<Result<List<Video>>> getVideosResultLiveData() {
        return videosResultLiveData;
    }

    public LiveData<Result<List<Review>>> getReviewsResultLiveData() {
        return reviewsResultLiveData;
    }

    public Utils getUtils() {
        return utils;
    }

    private void subscribeToVideosDataSource() {
        compositeDisposable.add(
                dataSource.getMovieTrailers(movie.getId())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(videos -> videosResultLiveData.setValue(Result.success(videos)),
                                error -> videosResultLiveData.setValue(Result.error(error.getMessage())))
        );
    }

    private void subscribeToReviewsDataSource() {
        compositeDisposable.add(
                dataSource.getMovieReviews(movie.getId())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(reviews -> reviewsResultLiveData.setValue(Result.success(reviews)),
                                error -> reviewsResultLiveData.setValue(Result.error(error.getMessage())))
        );
    }

    public void addToFavourites() {
        ioScheduler.runOnThread(() -> dataSource.addToFavourite(movie));
    }

    public void deleteFromFavourites() {
        ioScheduler.runOnThread(() -> dataSource.deleteFromFavourite(movie));
    }

    public Single<Movie> getMovieFromFavourites() {
        return dataSource.getFavouriteMovie(movie.getId());
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.dispose();
    }

}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/detail/DetailViewModel.kt
package com.example.ruslanyussupov.popularmovies.detail

import androidx.lifecycle.*
import androidx.paging.PagedList

import com.example.ruslanyussupov.popularmovies.App
import com.example.ruslanyussupov.popularmovies.data.DataSource
import com.example.ruslanyussupov.popularmovies.data.DataSource.Filter
import com.example.ruslanyussupov.popularmovies.data.Listing
import com.example.ruslanyussupov.popularmovies.data.model.Movie
import com.example.ruslanyussupov.popularmovies.data.model.Review
import com.example.ruslanyussupov.popularmovies.data.model.Video

import javax.inject.Inject

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber

class DetailViewModel : ViewModel() {

    @Inject
    internal lateinit var dataSource: DataSource

    lateinit var movie: Movie
    lateinit var filter: Filter

    private var markDelete = false

    private val _isFavorite by lazy {
        val result = MutableLiveData<Boolean>()
        viewModelScope.launch(Dispatchers.IO) {
            val state = dataSource.getFavourite(movie.id) != null
            result.postValue(state)
        }
        result
    }

    private val videosListing: LiveData<Listing<List<Video>>> by lazy {
        val result = MutableLiveData<Listing<List<Video>>>()
        viewModelScope.launch(Dispatchers.IO) {
            val listing = dataSource.getMovieVideos(filter, movie.id)
            result.postValue(listing)
        }
        result
    }

    private val reviewsListing: LiveData<Listing<PagedList<Review>>> by lazy {
        val result = MutableLiveData<Listing<PagedList<Review>>>()
        viewModelScope.launch(Dispatchers.IO) {
            val listing = dataSource.getMovieReviews(filter, movie.id)
            result.postValue(listing)
        }
        result
    }

    init {
        App.component?.inject(this)
    }

    fun setIsFavourite(state: Boolean) {
        if (state == _isFavorite.value) {
            return
        }
        viewModelScope.launch(Dispatchers.IO) {
            if (state) {
                addToFavourites()
            } else {
                deleteFromFavourites()
            }
            _isFavorite.postValue(state)
        }
    }

    fun videos() = Transformations.switchMap(videosListing) { it.data }

    fun videosNetworkState() = Transformations.switchMap(videosListing) { it.data }

    fun videosRefreshState() = Transformations.switchMap(videosListing) { it.refreshState }

    fun reviews() = Transformations.switchMap(reviewsListing) { it.data }

    fun reviewsNetworkState() = Transformations.switchMap(reviewsListing) { it.networkState }

    fun reviewsRefreshState() = Transformations.switchMap(reviewsListing) { it.refreshState }

    fun isFavorite() = _isFavorite

    fun retryVideos() {
        Timber.d("retryVideos")
        videosListing.value?.retry?.invoke()
    }

    fun retryReviews() {
        Timber.d("retryReviews")
        videosListing.value?.retry?.invoke()
    }

    private suspend fun addToFavourites() {
        markDelete = false
        GlobalScope.launch(Dispatchers.IO) { dataSource.insertFavourite(movie) }

    }

    private fun deleteFromFavourites() {
        markDelete = true
    }

    override fun onCleared() {
        super.onCleared()
        if (markDelete) {
            GlobalScope.launch(Dispatchers.IO) { dataSource.deleteFavourite(movie) }
        }
    }

}
